File: apollo-runtime/src/main/java/com/apollographql/apollo/cache/normalized/NormalizedCacheFactory.java
package com.apollographql.apollo.cache.normalized;

import com.apollographql.apollo.response.CustomTypeAdapter;
import com.apollographql.apollo.api.ScalarType;
import com.apollographql.apollo.api.internal.Function;
import com.apollographql.apollo.api.internal.Optional;

import org.jetbrains.annotations.NotNull;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

/**
 * A Factory used to construct an instance of a {@link NormalizedCache} configured with the custom scalar adapters set
 * in {@link com.apollographql.apollo.ApolloClient.Builder#addCustomTypeAdapter(ScalarType, CustomTypeAdapter)}.
 */
public abstract class NormalizedCacheFactory<T extends NormalizedCache> {
  private Optional<NormalizedCacheFactory> nextFactory = Optional.absent();

  /**
   * @param recordFieldAdapter A {@link RecordFieldJsonAdapter} configured with the custom scalar adapters set in {@link
   *                           com.apollographql.apollo.ApolloClient.Builder#addCustomTypeAdapter(ScalarType,
   *                           CustomTypeAdapter)}.
   * @return An implementation of {@link NormalizedCache}.
   */
  public abstract T create(RecordFieldJsonAdapter recordFieldAdapter);

  public final NormalizedCache createChain(final RecordFieldJsonAdapter recordFieldAdapter) {
    if (nextFactory.isPresent()) {
      return create(recordFieldAdapter)
          .chain(nextFactory.map(new Function<NormalizedCacheFactory, NormalizedCache>() {
            @NotNull @Override public NormalizedCache apply(@NotNull NormalizedCacheFactory factory) {
              return factory.createChain(recordFieldAdapter);
            }
          }).get());
    } else {
      return create(recordFieldAdapter);
    }
  }

  public final NormalizedCacheFactory<T> chain(@NotNull NormalizedCacheFactory factory) {
    checkNotNull(factory, "factory == null");

    NormalizedCacheFactory leafFactory = this;
    while (leafFactory.nextFactory.isPresent()) {
      leafFactory = (NormalizedCacheFactory) leafFactory.nextFactory.get();
    }
    leafFactory.nextFactory = Optional.of(factory);

    return this;
  }
}

File: apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/NormalizedCacheFactory.kt
package com.apollographql.apollo.cache.normalized

/**
 * A Factory used to construct an instance of a [NormalizedCache] configured with the custom scalar adapters set in
 * ApolloClient.Builder#addCustomScalarTypeAdapter(ScalarType, CustomScalarTypeAdapter).
 */
abstract class NormalizedCacheFactory<T : NormalizedCache> {

  private var nextFactory: NormalizedCacheFactory<out NormalizedCache>? = null

  /**
   * @param recordFieldAdapter A [RecordFieldJsonAdapter] configured with the custom scalar adapters set in
   * ApolloClient.Builder#addCustomScalarTypeAdapter(ScalarType, CustomScalarTypeAdapter).
   * @return An implementation of [NormalizedCache].
   */
  abstract fun create(recordFieldAdapter: RecordFieldJsonAdapter): T

  fun createChain(recordFieldAdapter: RecordFieldJsonAdapter): NormalizedCache {
    val nextFactory = nextFactory
    return if (nextFactory != null) {
      create(recordFieldAdapter).chain(nextFactory.createChain(recordFieldAdapter))
    } else {
      create(recordFieldAdapter)
    }
  }

  fun chain(factory: NormalizedCacheFactory<*>) = apply {
    var leafFactory: NormalizedCacheFactory<*> = this
    while (leafFactory.nextFactory != null) {
      leafFactory = leafFactory.nextFactory!!
    }
    leafFactory.nextFactory = factory
  }
}
